Add race/fav/type check functions

kei-gbf opened the pull request:
ref #296 

This patch add feature to check multiple fields for weapon skills.
This PR __will not add__ the race data update patch, (separate topic to game update)
It's now possible by String.includes method
e.g. akasha weapon checks fav like `"sowrd/dagger".includes("sword") => true`

This PR require wide range tests following:

----
### Checklist
Updated global_logic.js

- [x] baha skill refactoring
  - [x] bahaAT
  - [x] bahaATHP
  - [x] bahaHP
  - [x] bahaFUATHP
  - [x] bahaFUHP
- [x] cosmos skills
  - [x] __BUGFIX:__ wildcard had ignored skillname
  - [x] AT (type: attack)
  - [x] DF (type: defense)
  - [x] BL (type: balanse)
  - [x] PC (type: pecu)
- [x] akasha weapons
  - [x] Added wildcard check for 2nd skills
  - [x] fav: sword/dagger
  - [x] fav: wand/music (note: ex30% bcoz of akashaSensei)
  - [x] fav: axe/fist
  - [x] fav: bow/gun
  - [x] fav: spear/katana
- [x] Omega
  - [x] Fixed wildcard check for omega-gekijou (haisui graph)
  - [x] Fixed wildcard check for omega-kyousou (haisui graph)
- [x] magnaKenbu (fav: fist)
- [x] magnaJojutsu (fav: wand)
- [x] normalSeisyo (race: seisyo + Djeeta) ... ~~no need to include "unknown" or Djeeta?~~
- [x] magnaSeisyo (race: seisyo + Djeeta) ... include Djeeta, but not other unknown
- [x] normalBuff_doraf support (race: baha doraf __without wildcard check__) 
- [x] daBuff_fist support (__without wildcard check__) ... need verification info?

it's not replace all fav check,
LB zenith1 zenith2 and fav status bonus (20%) does not check wildcard.
I kept current code, not touch them.

----
Add skill_filter.js

This module provides check function for weapon skills

- check wildcard supports (now check "support", "support2", "support3")
  - we can modify the field keys if no need to check fields 2, 3
- check chara race
- check chara race with baha (unknown, primal)
- check chara fav
- check chara type

skill_filter.test.js has code sample and usage of those function.

kei-gbf commented:
for share Info for code reading.

----
### Array.some

it's useful utility function, widely use for common iteration tasks.
used to be in Lodash, we can use that standard in es6. it's primitive like a "for", "while"

```js
[true, true, true].every(v => v);  // true
[true, true, false].every(v => v);  // false
[true, true, false].some(v => v);  // true
[false, false, false].some(v => v);  // false
[1, 2, 3].some(num => num % 2 == 0); // can check array contains at least more than 1 even number
```

- Array.every/some is like a Python's all/any with function parameter
  - every -> if every elements returned true then true
  - some -> if at least one of element returned true then true

we can use this for fav/type/race/elements etc

----
### Array.includes, String.includes

```js
["A", "B", "C"].includes("A"); // => true
"ABC".includes("A"); // => true
```

it's the trick part, the check function can accept string or array of string.
both types has "includes" method. see Akasha weapon use both cases.

String.includes will check the multiple race
"havin/human".includes("human") => true

Exception/Implement Limitation, baha race function accept only array
because it add "unknown" and "seisyo" to __array__ so could not be string.

it was only one place ['doraf'] for normalBuff_doraf,
baha weapon races are from consts table, not add the parameter manually.

----
### WildCard checks

_withWildCardCheck function add the behavior on export timing.
note, it's does not change the original function itself, exported functions has the behavior.

- favContains(favs, values) ... does not check wildcard, values does not have support info
- favCharaContains(favs, chara) ... check wildcard, chara object has support info
  - notice the naming rule, raceCharaContains, typeCharaContains
    they take "chara" as parameter, then it's possible to check support fields.

- XXX: similar naming and checking wildcard implicitly.
    maybe confuse for code reading. if there are better idea we can rename.
    in unit tests, isTest/isTestOrWildcard ... to show it's explicitly check wildcard
    but in actual case, it's long and function parameter also different.

----

Q. you may worry, the most function are few lines, no much function call?

A. it's up to the script engine, how they optimize well. or acceptable cost in modern devices.
   it's depend on the implementation.

JavaScript is not functional programming language,
modern language specs picks a lot functional features (especially Array methods)
but run-time performance is not equal to them, they has better optimization.

I worry in here is the performance cost for old browsers.
I don't know how old browsers we should support.

A lot merits by this approach, small chunk of code are re-usable and test-able
I succeed refactor race/fav check code a lot, but
if performance issue, we should benchmark then back to hard-code for performance reason.

----

Q. why skill_filter.js file
A. I had support_filter.js with similar concept (not yet PR)

kei-gbf commented on src/global_logic.js:
```diff
@@ -1763,31 +1700,25 @@ module.exports.addSkilldataToTotals = function (totals, comb, arml, buff) {
-                            if (totals[key]["fav1"] == 'spear' || totals[key]["fav2"] == 'spear' ||
-                                totals[key]["fav1"] == 'katana' || totals[key]["fav2"] == 'katana') {
-                                    totals[key]["covenant"] = amount;
+                            if (favCharaContains('spear/katana', totals[key])) {
```
Add wildcard check for supplemental damage (akasha 2nd skill fav check)

kei-gbf commented on src/skill_filter.js:
```diff
@@ -0,0 +1,152 @@
+}
+
+
+// no favContains, it takes fav1,fav2 different interfaces.
```
outdated info, I will remove this comment

kei-gbf commented on src/global_logic.js:
```diff
@@ -2616,15 +2547,14 @@ module.exports.treatSupportAbility = function (totals, chara) {
-                                totals[key2]["race"] === "unknown" ||
-                                totals[key2]["race"] === "seisho") {
-                                totals[key2]["normalBuff"] += support.value
+                            // XXX: NO wildcard check?
```
If no need then I will remove this comment, before merge.

kei-gbf commented:
I had not integration test much, it requires more tests.
especially, wildcard + baha/cosmos/omega/akasha weapons.

kei-gbf commented:
This PR has now white-out error after updated branch. (maybe at resolve merge conflict)
missing "range" import from support_filter.js

kei-gbf commented:
The white-out error was fixed in branch HEAD. (it as error by merge update)
All features in this PR were already implemented.

I will use the PR comment checklist as confirmed the behavior. (when completed it's ok to merge)
anyone who has permission can check it, or report here what/how tested.

kei-gbf commented:
- [x] BUGFIX: wildcard had ignored Costmos skillname (Kaimu got cosmos effect wrong)
  - [x] Test Bug case: (current master) Kaimu always get Cosmos AT skill on Cosmos PC/DF/BL
  - [x] Test Buf fixed: (PR branch) Kaimu wildcard get correct Cosmos skills

cause the grouping of condition

```diff
- (skillname == 'cosmosAT' && totals[key]["type"] == "attack") || totals[key]["support"] === "wildcard"
+ skillname == 'cosmosAT' && (totals[key]["type"] == "attack" || totals[key]["support"] === "wildcard") 
```

is now the function check the wildcard internally.

```js
skillname == 'cosmosAT' && typeCharaContains("attack", totals[key])
```

- [x] コスモス武器スキルでのワイルドカードチェックの不具合修正
   原因: コスモスのスキル名チェックで AT のワイルドカードが常に true になる
  - [x] 現在のmasterブランチで、カイムのサポアビ(ワイルドカード)で
    コスモスのスキルが DF/BL/PC でも AT が適応されるのを確認（不具合の確認)
   - [x] PRのブランチでカイムに正常にコスモスのスキルが適応されるのを確認。（修正の確認）

kei-gbf commented:
- [ ] Added wildcard check for 2nd skills
  - [ ] Ok except Wand/Music

it's this PR bug, I will fix it soon

kagekiyo7 commented:
バハ武器のスキル適応については大丈夫そうです。
元からの不具合かも知れませんがスキル効果量の相違について

- TAはSlv15で8%
![image](https://user-images.githubusercontent.com/22173829/62423843-5034d200-b701-11e9-9c63-ede0a1b07006.png)
検証：http://hiroxgbf.blog.fc2.com/blog-entry-14.html

- この4つのHPスキル(フツルス)はSlv15で40%
![image](https://user-images.githubusercontent.com/22173829/62423854-78243580-b701-11e9-800d-f05bb79540f9.png)
![image](https://user-images.githubusercontent.com/22173829/62423858-82463400-b701-11e9-8b68-c5bfaf53ddc8.png)


効果量リスト：https://gbf-wiki.com/index.php?%A5%B9%A5%AD%A5%EB%B8%FA%B2%CC#xc5c7e99

kagekiyo7 commented:
星晶獣も種族不明と同様に無条件で効果を受けますが適応されていない？
![image](https://user-images.githubusercontent.com/22173829/62423984-4dd37780-b703-11e9-9f1d-fa489a19be81.png)

kagekiyo7 commented:
また、スキル名が**完全一致**の場合重複しないという面倒な仕様が…
![image](https://user-images.githubusercontent.com/22173829/62424056-73ad4c00-b704-11e9-8f83-ec0b0d6bb7af.png)
- バハ拳スキル
![image](https://user-images.githubusercontent.com/22173829/62424092-06e68180-b705-11e9-9f9a-5f97da65da5c.png)

kei-gbf commented:
> 効果量上限やスキル枠に関して

この PR では、種族チェックのみに専念するので、
上限チェックや枠重複に関しては別 ISSUE/PR でお願いします。

> 星晶獣も種族不明と同様に無条件で効果を受けますが適応されていない？

星晶獣を表すトークンの typo でした、seisyo -> seisho すぐに修正します
コードのテストではバハの種族チェックに星晶獣も含んでいるのを確認。
原因: 想定する実データが seisho だった typo。

kei-gbf commented:
関連で、通常星晶、方陣星晶のスキルでも同様の typo がありました。(但し、スキル名では Seisyou)

kei-gbf commented:
- [x] Kaimu with change element to "fire" or "water"
  - [x] magnaKenbu (fav: fist) "fire"
  - [x] magnaJojutsu (fav: wand) "water"

Currently, no actual cases Kaimu(wildcard) get those skills,
because of the element mismatch. However it's checked for the potential.

- [ ] normalSeisyo / magnaSeisyo

no earth Seisyo skill as well. when try to check

White-out error was found on previous the commit, it's missing import.
this never happen in current weapons set, but potential error. I will fix this soon.

kei-gbf commented:
- [x] normalBuff_doraf support (race: baha doraf without wildcard check)

fixed support ability tooltip text, it missing to mention Primal in ja/zh locales.

it contains Unknown and Seisyo like baha race checks,
but currently it's out of Wildcard (the text does not mention for support, only weapon skill)
this require verification, but for now it's expected behavior.

if we need wildcard check there then, code would be

```diff
- if (bahaRaceContains(["doraf"], totals[key2]["race"])) {
+ if (raceCharaContains(["doraf", "seisho". "unknown"], totals[key2])) {
```

----
Known bugs (won't fix in this PR)

```js
if (totals[key].isConsideredInAverage) {
    for (var key2 in totals) {
        // Draph, Unknown and Primal characters only (Fortified Vigor)
        if (bahaRaceContains(["doraf"], totals[key2]["race"])) {
            totals[key2]["normalBuff"] += support.value;
        }
    }
} else {
    // Calculate yourself only if you do not put it in the average
    totals[key]["normalBuff"] += support.value;  // XXX: Here does not check own race
}
```

this is not actual problem in general use case.
just report as potential bug, if the character *changed* the support ability but the character 
race is not any of them. she/he will get the buff in included *OFF* case.

I have an idea to improve those clone code by isConsideredInAverage flag,
support_filter.js provide utility functions to remove those clone codes.
that would be another PR.

kei-gbf commented:
## Tests for Baha weapons race checks

### Setup
- Characters
  - Djeeta (Unknown)
  - human
  - Erune
  - Doraf
  - Havin
  - Primal

### Baha race checks
- Weapon
  - [ ] bahaAT
  - [ ] bahaHP
  - [ ] bahaFUATHP
  - [ ] bahaFUHP
  - [ ] bahaFUHP

Insert any baha weapons, and check the skill applies to correct races.

### Wildcard checks

Baha race check internally has two race tables and
two functions provided `bahaRaceCharaContains` and `bahaFURaceCharaContains`
it's enough test just see wildcard check works for those two functions.
more wildcard test cases were in unit tests.


----
@Groovman 

This would be the last tasks in this PR, I will finish soon.
Can you add #296 update later?

This PR will extend the race check accepts multiple token like 'human/havin'
but I am not yet sure how race is treated in python json data converter.

kei-gbf commented:
Baha race checks, current status is: I got white-out error in some cases
maybe wrong refactoring had changed the logic.

race check logic should not move to the top-level of if-else block.

kei-gbf commented:
Done, this PR requires re-review it's changed after the first Approved.
fixes for Seisyou and Baha

----
Summary

This PR contains following fixes

- BUGFIX: Cosmos wildcard had ignored skillname
- Added wildcard check for Akasha 2nd skills
- Fixed race check for normal/magna Seisyou:  include Djeeta, but not other unknown
- Fixed wildcard check for omega-gekijou/kyousou (haisui graph)

require verification

- [ ] Wildcard is ignored in support ability, it's for now an expected behavior.
  as the game text mention weapon skill only.
  but there are support ability which implicitly include Primal race like baha weapon.
  so we need the verification.

----
主な変更内容: 種族/得意武器/タイプで複数の値を取れるようにする

#296 で複数属性を持つキャラが出て、従来の方法での拡張では対応が難しくなったため
キャラの種族や得意武器を判別するコードを全体的に整理しました。

このPRには以下の、不具合修正を含みます
- ワイルドカードでコスモスのAT以外のタイプが正常に適応されていなかった問題を修正
- スキルの星晶において、主人公が含まれていなかったのを修正
  （主人公以外の種族不明は含まれない）
- 虚空武器の第二スキルにワイルドカード適応
- 渾身背水グラフでのオメガ武器にワイルドカード適応
- サポート1でしかワイルドカードが有効にならなかった問題の修正

要検証 (このPRでは対応しない)
- [ ] サポートアビリティもワイルドカードの対象になるかどうか。
     テキストには武器スキルにしか言及されていないが、
     サポアビにもバハ武器と同じように暗黙的に星晶獣を含むものもある為、要検証。

kagekiyo7 commented:
雑にテストしましたが概ね問題ないと思います。
目の届きにくい箇所の修正までありがとうございました！

----
> スキルの星晶において、主人公が含まれていなかったのを修正
> （主人公以外の種族不明は含まれない）

検証ソース補足
https://twitter.com/KST33166653/status/1054985425989836801

kei-gbf commented:
![akasha](https://user-images.githubusercontent.com/45387555/63497326-7b3f6400-c4fe-11e9-8b1c-8f1ad0a26cf8.PNG)

report the related topic
it's an official game update fix 2019/08/22, Wildcard missing at akasha-axe skill

Groovman commented:
Complete now.
